sdks/java/core/src/main/java/org/apache/beam/sdk/runners/TransformHierarchy.java

```diff
@@ -462,30 +463,41 @@ void replaceOutputs(Map<PValue, ReplacementOutput> originalToReplacement) {
      * <p>Provides an ordered visit of the input values, the primitive transform (or child nodes for
      * composite transforms), then the output values.
      */
-    private void visit(PipelineVisitor visitor, Set<PValue> visitedValues) {
+    private void visit(PipelineVisitor visitor, Set<PValue> visitedValues, Set<Node> visitedNodes) {
+      if (visitedNodes.contains(this)) {
+        return;
+      }
+
       if (!finishedSpecifying) {
         finishSpecifying();
       }
 
       if (!isRootNode()) {
         // Visit inputs.
         for (PValue inputValue : inputs.values()) {
+          Node valueProducer = getProducer(inputValue);
+          if (!visitedNodes.contains(valueProducer)) {
```

Comment:
This is actually redundant since you check on the first line of `visit`. This whole routine might be refactored to isolate the  "already visited" checks to `visit` and `visitValue` with `actuallyVisit` and `actuallyVisitValue` that are more straight-line code. I'd just get this in and move forwards now, though.